WIP: Helm Client

spriadka opened the pull request:
Please make sure your PR meets the following requirements:
- [x] Pull Request contains a description of the changes
- [x ] Pull Request does not include fixes for multiple issues/topics
- [x] Code is formatted, imports ordered, code compiles and tests are passing
- [ ] Code is self-descriptive and/or documented

This is a work in progress PR enabling support for Helm Client

mchoma commented on core/src/main/java/cz/xtf/core/openshift/HelmClients.java:
```diff
@@ -0,0 +1,108 @@
+
+public class HelmClients {
+
+    private static final String HELM_CLIENTS_URL = "https://mirror.openshift.com/pub/openshift-v4/clients/helm/latest";
```
I believe we would need `latest` to be parametrized. `latest` is good default, but like we talked before there could be situation like OCP EUS, which will need we choose different version.

spriadka replied:
Fixed with e82acc779300f8a3c2130f7e031820ea6c71d9b5

mchoma commented:
Overal I think this is good. What is important to mention I haven't found any solid java API to be used, so seems we are really left to  handle helm by calling cli command. @spriadka did you considered java API as well?

mchoma reviewed (commented):
> Overal I think this is good. What is important to mention I haven't found any solid java API to be used, so seems we are really left to handle helm by calling cli command. @spriadka did you considered java API as well?

Although I came to this idea [1]. Tiller (helm server side) is protocol buffer and therefore could be generated with [2]. I think this is  worth to investigate.

[1] https://lairdnelson.wordpress.com/2017/06/14/helm-from-java/
[2] https://github.com/grpc/grpc-java

mnovak1 commented on core/src/main/java/cz/xtf/core/openshift/HelmBinary.java:
```diff
@@ -0,0 +1,54 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class HelmBinary {
```
At the first look, this looks like duplication of:
https://github.com/xtf-cz/xtf/blob/master/core/src/main/java/cz/xtf/core/openshift/OpenShiftBinary.java

together with `CLIUtils` class above.

Could be `OpenShiftBinary` used instead? Or delegated?

mchoma replied:
At least here is `helmConfigPath` which differentiate. Of course some code can be shared but conceptually I think it is good idea to have separate classes for OpenShiftBinary (`oc` command) and HelmBinary (`helm` command)

mnovak1 replied:
Right, everything what is specific for `HelmBinary` can be in this class. This is about checking that code is not duplicated in `OpenShiftBinary`.

spriadka commented on core/src/main/java/cz/xtf/core/openshift/HelmClients.java:
```diff
@@ -0,0 +1,109 @@
+
+public class HelmClients {
+
+    private static final String HELM_CLIENTS_URL = "https://mirror.openshift.com/pub/openshift-v4/clients/helm/latest";
```
Tiller support has been deprecated so we depend on wrapping the CLI calls. I've changed the URL for helm binary download from `mirror.openshift.com` take a look

mchoma replied:
> Tiller support has been deprecated so we depend on wrapping the CLI calls.

Sorry, I am failing to understand how you mean in this context 

> I've changed the URL for helm binary download from mirror.openshift.com take a look

I still think we should parametrize url and choose one default.

spriadka replied:
So we'll end up with parametrized downloads URL and helm version? (Just asking to make it clear, the implementation is rather simple with this one)

mchoma replied:
Yes, I think it is enough. That will provide us flexibility to user to change what he need.

I think it is ok to default to `https://mirror.openshift.com/pub/openshift-v4/clients/helm/` and `latest` version. But still have way to download from `helm.sh` and specific version.

spriadka commented on core/src/main/java/cz/xtf/core/openshift/HelmClients.java:
```diff
@@ -0,0 +1,109 @@
+
+    private static HelmBinary getBinary(String token, String username, String password, String kubeconfig,
+            String namespace) {
+        OpenShiftBinary openShiftBinary = OpenShifts.getBinary(token, username, password, kubeconfig, namespace);
```
`HelmBinary` is depending on `oc.config` that is being downloaded with the `oc.binary`.  The `OpenSjifts.getBinary()` calls downloads the binary, config and performs login on that config, so I found it convenient, as we'll stay in the same context as with OpenShift binary

spriadka replied:
I guess we can have a workaround by passing env variables to helm binary. This may result in not requiring `oc.config` at all (https://helm.sh/docs/helm/helm/)

mchoma replied:
I am not sure I follow. Maybe some chat or short call will help me. Are we trying to solve som real problem?

> I found it convenient, as we'll stay in the same context as with OpenShift binary

So is it about reusing `oc.config` path? Is solution for that just expose constant for `oc.config` and use it on both places? 

> I think we can do this with HELM_KUBETOKEN and HELM_KUBEASUSER variables

Or is it about sharing token? But we already have this parameters as input `String token, String username`

spriadka replied:
We can 1) reuse `oc.config`, but in order to do that, we need to download it first. AFAIK only `OpenShiftBinary` uses this config and `OpenShift` is usually configured to use `OpenShiftConfig` internally (fabric8 API). `OpenShiftBinary` will download it with the binary itself, that's why I've decided to bundle `helm` binary with `oc` binary. 
2) If we don't want to re-use `oc.config` (or use it at all), we may not need to download it at all, and we can provide the environment variables to `ProcessBuilder`

AFAIK `oc.config` has `context` entries which point to user activity within namespace, so using `oc.config` can be fully replaced using 2)

mchoma replied:
`reuse oc.config, but in order to do that, we need to download it first.` download from where? I thought when we talk about `oc.config` we talk about `~/.kube/config` local machine file filled with entries based on how `oc login` is used on that machine.

Is `oc.config` `~/.kube/config` ?

mchoma replied:
Or in other words my understanding is we are taling here about 

```
$KUBECONFIG	set an alternative Kubernetes configuration file (default "~/.kube/config")
```
or
```
--kubeconfig string           path to the kubeconfig file
```
Described in https://helm.sh/docs/helm/helm/ . Am I missing something?

spriadka replied:
I was talking about case of `OpenShifts.getBinary()` (https://github.com/xtf-cz/xtf/blob/master/core/src/main/java/cz/xtf/core/openshift/OpenShifts.java#L122) method, which will create empty `oc.config` in temporary directory, so the whole interaction of `OpenShitBinary` when running TS will not affect user's local `~/.kube/config`.

mchoma replied:
Ok, so I am starting to understand it. `oc.config` is internal way of OpenShiftBinary how to keep authentication context. So you are basically delegating login logic to OpenShiftBinary. I understand that was easiest and most straighforward way.

Now I do not now if this is elegant or hack :) I tend to think it is hack :)

Conceptually I think HelmBinary and OpenShiftBinary should be independent.

I know there were cases where mystoriously in OpenShiftBinary authentication context was switching and I rather explicitely made .login() to ensure corect user is used with OpenShiftBinary. Probably it was because internal oc.config was changed.

In other words can it happen I `OpenShiftBinary.login()` with one user and I will switch user with `getHelmBinary()` call? I think that is dangerous.

But I still admit I do not see situation clear.

mchoma replied:
@jbliznak this is discussion, why I am not sure if we want this in in this form.

spriadka replied:
@mchoma I agree on decoupling `OpenShiftBinary` and `HelmBinary`. I tried getting rid of `oc.config` and use env variables only (7e62dadf83252aa273b763b788cd25bdf5ad025e) but looks like, in addition to `HELM_KUBEASUSER`, `HELM_KUBETOKEN` and `HELM_KUBEAPISERVER`, we need to set `HELM_KUBECAFILE` env variable as well in order to establish secure connection with the K8S server.  This brings user rather complications into the usage of `HelmBinary` IMHO.

spriadka replied:
`oc.config` is AFAIK temporary `kubeconfig` that in our case and our method calls, will be downloaded for each instance of `HelmBinary`. `OpenShiftBinary.login()` is being called (https://github.com/xtf-cz/xtf/blob/master/core/src/main/java/cz/xtf/core/openshift/OpenShifts.java#L129) everytime the `OpenShiftBinary` would is downloaded and every call of `OpenShifts.getBinary()` downloads new `OpenShiftBinary` instance so I don't think there would be any OpenShift context interference

mchoma replied:
We have agreed on call. It seems it is not possible fully decouple OpenShiftBinary and HelmBinary.  The reason for that is HelmBinary does not have same capabilities [1] when it comes to authenticating to OpenShift as `oc`. Helm can authenticate with 
- kubeconfig - but we need to gain it with `oc`
- token - we can use it from test.properties, but still we do not have support for trustall and successfully create connection with SSL

[1] https://helm.sh/docs/helm/helm/#options

spriadka commented on core/src/main/java/cz/xtf/core/openshift/HelmClients.java:
```diff
@@ -0,0 +1,109 @@
+    private static HelmBinary getBinary(String token, String username, String password, String kubeconfig,
+            String namespace) {
+        OpenShiftBinary openShiftBinary = OpenShifts.getBinary(token, username, password, kubeconfig, namespace);
+        return new HelmBinary(downloadHelmBinary(), openShiftBinary.getOcConfigPath());
```
+1

spriadka commented on core/src/main/java/cz/xtf/core/image/Product.java:
```diff
@@ -29,6 +29,22 @@ public String templatesBranch() {
+        return resolveDefaultingProperty("helm.charts.branch");
+    }
+
+    public String helmChartsLocation() {
```
It's just parametrized directory within repository for helm charts (e.g https://github.com/wildfly/wildfly-charts/tree/main/charts). This just provides option to parameterized the location for charts.

mchoma replied:
So is this property required? If I do not specify `helm.charts.location` will it default to `charts` ?

spriadka replied:
You're correct that the property is optional, however the defaulting behaviour is managed by test suites (or other software using XTF), in our use case, it will default to `charts`

mchoma replied:
Ok, I agree default in xtf based on EAP knowledge could miss in other products. Not sure there is some convention among different products how helm repo is structured. I believe not.

Setting default in testsuite seems correct to me.

spriadka commented on core/src/main/java/cz/xtf/core/image/Product.java:
```diff
@@ -29,6 +29,22 @@ public String templatesBranch() {
+    }
+
+    public String helmChartExamplesLocation() {
+        return resolveDefaultingProperty("helm.charts.examples.location");
```
Simillarly to the charts parametrization, let's say we may have `examples-wildfly23, examples-wildfly24`. It really is just. about adding possibility. to adjust these properties to the repository layout

mchoma replied:
So is this property optional? If I do not specify `helm.charts.examples.location` will it default to `examples` ?

mchoma commented on core/src/main/java/cz/xtf/core/openshift/OpenShifts.java:
```diff
@@ -117,7 +117,7 @@ public static OpenShiftBinary adminBinary(String namespace) {
                 OpenShiftConfig.adminKubeconfig(), namespace);
     }
 
-    private static OpenShiftBinary getBinary(String token, String username, String password, String kubeconfig,
```
This is not necessary after deleting dependency between HelmBinary and OpenShiftBinary?

spriadka replied:
I have made quite some changes, would you have another look if this thread can be resolved?

mchoma replied:
I see this method was resolved from OpenShifts. Resolving.

mchoma commented on core/src/main/java/cz/xtf/core/openshift/HelmClients.java:
```diff
@@ -0,0 +1,107 @@
+package cz.xtf.core.openshift;
```
What about putting helm specific things into own package, e.g. `cz.xtf.core.helm`?

After removing dependency between HelmBinary and OpenShiftBinary there is no reason to be part of `cz.xtf.core.openshift` ?

mchoma commented on core/src/main/java/cz/xtf/core/openshift/HelmBinary.java:
```diff
@@ -0,0 +1,56 @@
+package cz.xtf.core.openshift;
```
What about putting helm specific things into own package, e.g. cz.xtf.core.helm?

After removing dependency between HelmBinary and OpenShiftBinary there is no reason to be part of cz.xtf.core.openshift ?

mchoma commented on core/src/main/java/cz/xtf/core/openshift/HelmBinary.java:
```diff
@@ -0,0 +1,56 @@
+        environmentVariables.put("HELM_CONFIG_HOME", helmConfigPath);
+        environmentVariables.put("HELM_KUBEAPISERVER", OpenShiftConfig.url());
+        environmentVariables.put("HELM_KUBEASUSER", kubeUsername);
+        environmentVariables.put("HELM_KUBETOKEN", kubeToken);
```
Token can be empty. Does it work as expected in that case?

mchoma commented on core/src/main/java/cz/xtf/core/openshift/HelmBinary.java:
```diff
@@ -0,0 +1,56 @@
+        Map<String, String> environmentVariables = new HashMap<>();
+        environmentVariables.put("HELM_CONFIG_HOME", helmConfigPath);
+        environmentVariables.put("HELM_KUBEAPISERVER", OpenShiftConfig.url());
+        environmentVariables.put("HELM_KUBEASUSER", kubeUsername);
```
Username (`OpenShiftConfig.adminUsername()` or `OpenShiftConfig.masterUsername()`) can be empty. Does it work as expected in that case?

mchoma replied:
I do not understand why password is missing here.

For OpenShift Client authentication is done either by token or by username/password. If token is used ofr authentication username is not necessary.

mchoma commented on core/src/main/java/cz/xtf/core/openshift/HelmBinary.java:
```diff
@@ -0,0 +1,56 @@
+        this.namespace = namespace;
+    }
+
+    public String execute(String... args) {
```
So we are authenticating with each `execute` call? I am not saying that is bad, I just want to understand it.

For OpenShiftBinary, there was that idea of one login and subsequent calls in that authentication context.

I do not have preference right now. But did you compared both approaches and find authenticate on each call better?

mchoma commented on core/src/main/java/cz/xtf/core/openshift/HelmBinary.java:
```diff
@@ -0,0 +1,56 @@
+
+    public String execute(String... args) {
+        Map<String, String> environmentVariables = new HashMap<>();
+        environmentVariables.put("HELM_CONFIG_HOME", helmConfigPath);
```
My feeling is we should use directly command parameters if this is possible instead of env variables.

Environmnet variables can be used to tweak behaviour from outside. but I do not have strong opinion.

mchoma replied:
Ok

fabiobrz commented:
Hi @spriadka - Is there any chance to evaluate whether supporting Helm can be done without depending on (yet) another binary?
I haven't had any look at it but at a quick search I stumbled upon [this](https://github.com/microbean/microbean-helm), just as an example.

mchoma commented:
> Hi @spriadka - Is there any chance to evaluate whether supporting Helm can be done without depending on (yet) another binary? I haven't had any look at it but at a quick search I stumbled upon [this](https://github.com/microbean/microbean-helm), just as an example.

Hi @fabiobrz. We did consider java API, look [1] or [2]. We also found project you are mentioning, problem is it does not seem active anymore they stopped supporting version 2.x, wheares we need 3.x. We didn't found any fitting project back then. Seemed we would need to start writing something ourselves, which would be time consuming.

[1] https://github.com/xtf-cz/xtf/pull/422#issuecomment-815886464
[2] https://github.com/xtf-cz/xtf/pull/422#pullrequestreview-631318272

fabiobrz commented:
> We did consider java API, look [1] or [2]. We also found project you are mentioning, problem is it does not seem active anymore they stopped supporting version 2.x, wheares we need 3.x. We didn't found any fitting project back then. Seemed we would need to start writing something ourselves, which would be time consuming.

Thanks for your feedback @mchoma and sorry for not noticing the previous discussion about the topic.
I'll have a look around again, searching for alternatives (maybe a fork which carried the original project on?) because although ai perfectly understand the capacity issues, depending on binaries seems to me really the last resort choice.

mnovak1 commented:
@spriadka Hi, could you update discussions in this PR, please? We would like to release new version of XTF with this merged. Thanks.

simkam commented on core/src/main/java/cz/xtf/core/openshift/OpenShiftBinaryManagerFactory.java:
```diff
@@ -4,7 +4,7 @@
 import java.util.stream.Stream;
 
-enum OpenShiftBinaryManagerFactory {
+public enum OpenShiftBinaryManagerFactory {
```
why this change?

simkam commented on core/src/main/java/cz/xtf/core/openshift/OpenShiftBinaryManager.java:
```diff
@@ -40,7 +40,7 @@ OpenShiftBinary adminBinary(String namespace) {
     }
 
-    private OpenShiftBinary getBinary(String token, String username, String password, String kubeconfig,
+    public OpenShiftBinary getBinary(String token, String username, String password, String kubeconfig,
```
why this change?

simkam commented on core/src/main/java/cz/xtf/core/helm/HelmClients.java:
```diff
@@ -0,0 +1,108 @@
+import cz.xtf.core.http.Https;
+import cz.xtf.core.openshift.CLIUtils;
+
+public class HelmClients {
```
Could you please update this to follow the same pattern as oc binary.

public OpenShiftBinaryManagerFactory + package-private OpenShiftBinaryManager + path resolver(s)

it would be thread-safe and testable (also please add tests)

mchoma commented on README.md:
```diff
@@ -247,6 +247,19 @@ JUnit5 module provides a number of extensions and listeners designed to easy up
 ## Releasing XTF
 Have a look to the [release documentation](RELEASE.md) to learn about the process that defines how to release XTF to 
 the community.
+JUnit5 module provides number of extensions and listeners designed to easy up OpenShift images test management. See [JUnit5](https://github.com/xtf-cz/xtf/blob/master/core/src/main/java/cz/xtf/core/waiting/SimpleWaiter.java) for more informations. 
```
Some copy paste error? I see same text above.

mchoma commented on core/src/main/java/cz/xtf/core/config/HelmConfig.java:
```diff
@@ -0,0 +1,26 @@
+    public static final String HELM_BINARY_CACHE_PATH = "xtf.helm.binary.cache.path";
+    public static final String HELM_BINARY_CACHE_ENABLED = "xtf.helm.binary.cache.enabled";
+    public static final String HELM_BINARY_PATH = "xtf.helm.binary.path";
+    public static final String HELM_CLIENT_VERSION = "xtf.helm.client.version";
```
So for consistency can you provide also method for `HELM_CLIENT_VERSION` and use that method in rest of MR?

mchoma commented on core/src/main/java/cz/xtf/core/helm/HelmBinary.java:
```diff
@@ -0,0 +1,57 @@
+        this.namespace = namespace;
+    }
+
+    public String execute(String... args) {
```
We talked about some fluent api would be nice for most frequent commands. That is out of scope of this MR?

mchoma commented:
I will be off one week. I do not have any important objections which could block this MR. I understand main exposed API for users are HelmBinary, HelmClients, HelmConfig, which seems robust enough to start with it. We can add next features as needed in future. From my perspective we can move on with this.

simkam commented on core/src/main/java/cz/xtf/core/helm/HelmClients.java:
```diff
@@ -0,0 +1,20 @@
+        return getBinary(OpenShiftConfig.masterToken(), OpenShiftConfig.masterUsername(), OpenShiftConfig.namespace());
+    }
+
+    private static HelmBinary getBinary(String token, String username, String namespace) {
```
this should be probably moved to HelmBinaryManager

simkam commented on core/src/main/java/cz/xtf/core/helm/HelmClients.java:
```diff
@@ -0,0 +1,20 @@
+public class HelmClients {
+
+    public static HelmBinary adminBinary() {
+        return getBinary(OpenShiftConfig.adminToken(), OpenShiftConfig.adminUsername(), OpenShiftConfig.namespace());
```
these methods should probably delegate to instantiable HelmBinaryManager

HelmBinaryManagerFactory.INSTANCE.getHelmBinaryManager().adminBinary()

simkam commented on core/src/main/java/cz/xtf/core/helm/ClusterVersionHelmBinaryPathResolver.java:
```diff
@@ -0,0 +1,138 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ClusterVersionHelmBinaryPathResolver implements HelmBinaryPathResolver {
```
package-private

simkam commented on core/src/main/java/cz/xtf/core/helm/ConfiguredPathHelmBinaryResolver.java:
```diff
@@ -0,0 +1,10 @@
+
+import cz.xtf.core.config.HelmConfig;
+
+public class ConfiguredPathHelmBinaryResolver implements HelmBinaryPathResolver {
```
package-private

simkam commented on core/src/main/java/cz/xtf/core/openshift/OpenShiftBinaryManager.java:
```diff
@@ -40,7 +40,7 @@ OpenShiftBinary adminBinary(String namespace) {
     }
 
-    private OpenShiftBinary getBinary(String token, String username, String password, String kubeconfig,
+    OpenShiftBinary getBinary(String token, String username, String password, String kubeconfig,
```
why this change?

simkam reviewed (changes requested):
tests don't pass for me locally so I didn't check them for now, but I might have messed rebase.

please rebase and squash commits where it makes sense

spriadka commented:
Thanks @simkam for resolving the threads, I got the CI passing (the problem was different child filename of the helm archives depending on the platform). I've also rebased and squashed commits

spriadka commented:
@simkam would you have another look at `getHelmClientUrlBasedOnConfiguredHelmVersion` and the `resolve` methods? Please resolve the threads if you're satisfied with the proposed solutions

simkam commented on core/src/main/java/cz/xtf/core/helm/ClusterVersionHelmBinaryPathResolver.java:
```diff
@@ -0,0 +1,132 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+class ClusterVersionHelmBinaryPathResolver implements HelmBinaryPathResolver {
```
can you please rename this and related tests? Perhaps something like `ConfiguredVersionHelmBinaryPathResolver`

simkam commented on core/src/main/java/cz/xtf/core/helm/ClusterVersionHelmBinaryPathResolver.java:
```diff
@@ -0,0 +1,132 @@
+    public String resolve() {
+        final ClusterVersionInfo clusterVersionInfo = ClusterVersionInfoFactory.INSTANCE.getClusterVersionInfo();
+        if (!clusterVersionInfo.getOpenshiftVersion().startsWith("4")) {
+            log.info(
```
```suggestion
            log.warn(
```

simkam reviewed (changes requested):
Thanks, last two minor comments and we can merge this.

simkam reviewed (approved):
LGTM, but please squash commits
